src/compiler/scala/tools/nsc/Global.scala

```diff
@@ -1637,6 +1639,7 @@ class Global(var currentSettings: Settings, reporter0: LegacyReporter)
 
     /** Reset package class to state at typer (not sure what this is needed for?)
      */
+    @tailrec
```

Comment:
Since we are at it, can we make this `private[this]`. In fact, can we backport https://github.com/scala/scala/pull/7127 to `2.12.x`?

Reply:
Someone could volunteer for private this patrol. Also, if they're not going to make private mean private this in scala 3, then `-Xlint:private` should make recommendations and `-Xlint-fix` should scalafix them.

Reply:
I'd like to keep this PR focussed on the one cross-cutting improvement.

In general, 2.12.x development should start winding down, so backports need to pay their way